Implement etcd-backup-restore on OSS of Alicloud

minchaow opened the pull request:
**What this PR does / why we need it**:
Implement function of backing up and restoring etcd on OSS of Alicloud

**Special notes for your reviewer**:
1. When running "make verify", there was one failure testing case as follow:

------------------------------
• Failure [2.036 seconds]
Defrag Defragmentation [It] should defragment and reduce size of DB within time 
/Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/defrag_test.go:53

  Expected
      <int64>: 258048
  to be <
      <int64>: 258048

  /Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/defrag_test.go:69

  Full Stack Trace
  	/Workspace/go/src/github.com/gardener/etcd-backup-restore/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:75 +0x305
  github.com/gardener/etcd-backup-restore/vendor/github.com/onsi/gomega/internal/assertion.(*Assertion).Should(0xc000938400, 0x1801ea0, 0xc0006fa570, 0x0, 0x0, 0x0, 0x1)
  	/Workspace/go/src/github.com/gardener/etcd-backup-restore/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:28 +0x11f
  github.com/gardener/etcd-backup-restore/pkg/etcdutil.glob..func1.1.2()
  	/Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/defrag_test.go:69 +0xc72
  github.com/gardener/etcd-backup-restore/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0002f0060, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
  	/Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/etcdutil_suite_test.go:42 +0xf7
  testing.tRunner(0xc0001f4500, 0x164df60)
  	/usr/local/go/src/testing/testing.go:827 +0x163
  created by testing.(*T).Run
  	/usr/local/go/src/testing/testing.go:878 +0x65a
  
------------------------------

However, when I ran on the latest master branch without my change, the same error was still happened. So it is confused me if it is my code causes this error.

2. The implement needs 3rd part SDK, so I appended the dependency into [[constraint]]  in the Gopkg.toml. When running "make revendor", not only aliyun SDK was added, but also some other dependencies were updated, which cause one problem as follow when to run "make verify":

panic: codecgen version mismatch: current: 8, need 10. Re-generate file: /Workspace/go/src/github.com/gardener/etcd-backup-restore/vendor/github.com/coreos/etcd/client/keys.generated.go

goroutine 1 [running]:
github.com/gardener/etcd-backup-restore/vendor/github.com/coreos/etcd/client.init.0()
	/Workspace/go/src/github.com/gardener/etcd-backup-restore/vendor/github.com/coreos/etcd/client/keys.generated.go:45 +0x144

Ginkgo ran 1 suite in 4.835755187s
Test Suite Failed
Makefile:61: recipe for target 'test' failed
make: *** [test] Error 1

The same error was also happened when running on the latest master branch.


**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator
-->
```noteworthy user
Add new cloud provider OSS (Alibaba Object Storage Service) support for etcd-backup-restore
```

swapnilgm commented:
@minchaow Thank you for the PR. We will review it soon.

Regarding test cases:
 All tests for master and recent PR are passing in CI/CD pipeline and locally on my machine as well. So there shouldn't be any issue.

Anyways, looking at 
> Defrag Defragmentation [It] should defragment and reduce size of DB within time
/Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/defrag_test.go:53``

It looks like DB defragmentation ran on already defragmented etcd. By any chance, any other process calling defragmentation on etcd? I'm not able to reproduce it on master.
Just for safer side, try to run the test in docker container. You can use something like below:
```
docker run --rm -it \
    -v ${PWD}:/master-version \
   -e GOPATH="/go" \
   -w /master-version golang bash -c "make verify"
``` 

> The implement needs 3rd part SDK, so I appended the dependency into [[constraint]] in the Gopkg.toml. When running "make revendor", not only aliyun SDK was added, but also some other dependencies were updated, which cause one problem as follow when to run "make verify":

This issue is similar to https://github.com/etcd-io/etcd/issues/8715 . Might have been triggered due to later release of go-codec library [1.1.2](https://github.com/ugorji/go/releases/tag/v1.1.2).
Till upstream fixes this, do the following.
Please add following line to `Gopkg.toml` at the end and run `make revendor` again. Make sure that `Gopkg.lock` reflects the codec version in use as 1.1.1.

```
[[override]]
name = "github.com/ugorji/go"
revision = "b4c50a2b199d93b13dc15e78929cfb23bfdf21ab"
```

swapnilgm commented on pkg/snapstore/oss_snapstore.go:
```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
```
s/2018/2019/g

minchaow replied:
done!

swapnilgm commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
```
s/2018/2019/g

minchaow replied:
done

minchaow commented:
> @minchaow Thank you for the PR. We will review it soon.
> 
> Regarding test cases:
> All tests for master and recent PR are passing in CI/CD pipeline and locally on my machine as well. So there shouldn't be any issue.
> 
> Anyways, looking at
> 
> > Defrag Defragmentation [It] should defragment and reduce size of DB within time
> > /Workspace/go/src/github.com/gardener/etcd-backup-restore/pkg/etcdutil/defrag_test.go:53``
> 
> It looks like DB defragmentation ran on already defragmented etcd. By any chance, any other process calling defragmentation on etcd? I'm not able to reproduce it on master.
> Just for safer side, try to run the test in docker container. You can use something like below:
> 
> ```
> docker run --rm -it \
>     -v ${PWD}:/master-version \
>    -e GOPATH="/go" \
>    -w /master-version golang bash -c "make verify"
> ```
> > The implement needs 3rd part SDK, so I appended the dependency into [[constraint]] in the Gopkg.toml. When running "make revendor", not only aliyun SDK was added, but also some other dependencies were updated, which cause one problem as follow when to run "make verify":
> 
> This issue is similar to [etcd-io/etcd#8715](https://github.com/etcd-io/etcd/issues/8715) . Might have been triggered due to later release of go-codec library [1.1.2](https://github.com/ugorji/go/releases/tag/v1.1.2).
> Till upstream fixes this, do the following.
> Please add following line to `Gopkg.toml` at the end and run `make revendor` again. Make sure that `Gopkg.lock` reflects the codec version in use as 1.1.1.
> 
> ```
> [[override]]
> name = "github.com/ugorji/go"
> revision = "b4c50a2b199d93b13dc15e78929cfb23bfdf21ab"
> ```

@swapnilgm Thank you very much. I will modify my code asap and run the UT cases again according to your suggestion.

Best regards!

minchaow commented:
@swapnilgm , thanks for your suggestions very much. UT has passed when running inside the container, and the workaround for my second issue is workable as well.
Also, I have changed my code according to your review comments and added license information.

minchaow commented:
@swapnilgm, this PR is important for enabling gardener on alicloud, so could you please help review the code, the sooner the better, many thanks for you help!  :)

swapnilgm commented:
@minchaow Thanks for addressing the changes. I can understand the importance of the PR. We will review it soon.

swapnilgm commented:
@shreyas-s-rao  Please review the PR. 
@minchaow We don't have access to Alicloud, so we can just theoretically review the PR and can't test it with actual OSS. If required, Shreyas might get back to you over slack or mail to check the demo with actual deployment.

swapnilgm commented:
Sorry my bad. Just received the access to AliCloud.

minchaow commented:
Hi @shreyas-s-rao, code has changed according to suggestions, please help review again. Thanks a lot.

shreyas-s-rao commented:
@minchaow After running `make revendor`, push the code along with all the new files that are added to the `vendor` directory, so that the OSS SDK gets added in-tree.

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	bucketName            string
+}
+
+func (m *mockOSSBucket) GetObject(objectKey string, options ...oss.Option) (io.ReadCloser, error) {
```
Add function description comment

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	return out, nil
+}
+
+func (m *mockOSSBucket) InitiateMultipartUpload(objectKey string, options ...oss.Option) (oss.InitiateMultipartUploadResult, error) {
```
See above comment

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	}, nil
+}
+
+func (m *mockOSSBucket) UploadPart(imur oss.InitiateMultipartUploadResult, reader io.Reader, partSize int64, partNumber int, options ...oss.Option) (oss.UploadPart, error) {
```
See above comment

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	}, nil
+}
+
+func (m *mockOSSBucket) CompleteMultipartUpload(imur oss.InitiateMultipartUploadResult, parts []oss.UploadPart, options ...oss.Option) (oss.CompleteMultipartUploadResult, error) {
```
See above comment

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	}, nil
+}
+
+func (m *mockOSSBucket) ListObjects(options ...oss.Option) (oss.ListObjectsResult, error) {
```
See above comment

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+	return out, nil
+}
+
+func (m *mockOSSBucket) DeleteObject(objectKey string) error {
```
See above comment

minchaow replied:
done

minchaow commented on pkg/snapstore/oss_snapstore.go:
```diff
@@ -0,0 +1,274 @@
+)
+
+// OSSBucket is an interface for oss.Bucket used in snapstore
+type OSSBucket interface {
```
Yes, OSS SDK doesn't provider the interface. We need rewrite one for mock test.

minchaow commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+		Bucket: m.bucketName,
+	}, nil
+}
+
```
done

minchaow commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,150 @@
+import (
+	"bytes"
+	"fmt"
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
```
done

minchaow commented on pkg/snapstore/snapstore_test.go:
```diff
@@ -87,6 +87,12 @@ var _ = Describe("Snapstore", func() {
 				objects: objectMap,
 				prefix:  prefix,
 			}),
+			"oss": NewOSSFromBucket(prefix, "/tmp", 5, &mockOSSBucket{
```
done

minchaow commented:
@swapnilgm, @shreyas-s-rao, comments addressed, and changes pushed. Please help review. Thanks.

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,163 @@
+
+// AbortMultipartUpload returns the result of aborting upload.
+func (m *mockOSSBucket) AbortMultipartUpload(imur oss.InitiateMultipartUploadResult, options ...oss.Option) error {
+	delete(m.objects, imur.UploadID)
```
s/m.objects/m.multiPartUploads/

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,163 @@
+func (m *mockOSSBucket) ListObjects(options ...oss.Option) (oss.ListObjectsResult, error) {
+	var contents []oss.ObjectProperties
+	for key := range m.objects {
+		tempobj := oss.ObjectProperties{
```
s/tempobj/tempObj/

minchaow replied:
done

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,163 @@
+		tempobj := oss.ObjectProperties{
+			Key: key,
+		}
+		contents = append(contents, tempobj)
```
s/tempobj/tempObj/

minchaow replied:
done

minchaow replied:
done

minchaow commented:
@shreyas-s-rao, comments addressed, and the commit pushed.

shreyas-s-rao commented on pkg/snapstore/oss_snapstore_test.go:
```diff
@@ -0,0 +1,163 @@
+		return oss.UploadPart{}, fmt.Errorf("file should not be null")
+	}
+
+	if m.multiPartUploads[imur.UploadID] == nil {
```
This line of code must be within the scope of the `m.multiPartUploadsMutex.Lock()`, or else it could lead to data race. Please move the `m.multiPartUploadsMutex.Lock()` above this line to fix it.

minchaow replied:
done

minchaow commented:
@shreyas-s-rao, comments addressed. Thanks for your so careful review! 👍 😄
